docker.go

```diff
@@ -1122,29 +1122,80 @@ func (p *DockerProvider) CreateContainer(ctx context.Context, req ContainerReque
 
 	req.LifecycleHooks = []ContainerLifecycleHooks{combineContainerHooks(defaultHooks, req.LifecycleHooks)}
 
-	err = req.creatingHook(ctx)
-	if err != nil {
-		return nil, err
+	if req.Reuse {
+		// Remove the SessionID label from the request, as we don't want Ryuk to control
+		// the container lifecycle in the case of reusing containers.
+		delete(req.Labels, core.LabelSessionID)
+	}
+
+	var resp container.CreateResponse
+	if req.Reuse {
+		// we must protect the reusability of the container in the case it's invoked
+		// in a parallel execution, via ParallelContainers or t.Parallel()
+		reuseContainerMx.Lock()
+		defer reuseContainerMx.Unlock()
+
+		// calculate the hash, and add the labels, just before creating the container
+		hash := req.hash()
+		req.Labels[core.LabelContainerHash] = fmt.Sprintf("%d", hash.Hash)
+		req.Labels[core.LabelCopiedFilesHash] = fmt.Sprintf("%d", hash.FilesHash)
+
+		// in the case different test programs are creating a container with the same hash,
+		// we must check if the container is already created. For that we wait up to 5 seconds
+		// for the container to be created. If the error means the container is not found, we
+		// can proceed with the creation of the container.
+		// This is needed because we need to synchronize the creation of the container across
+		// different test programs.
+		c, err := p.waitContainerCreationInTimeout(ctx, hash, 5*time.Second)
```

Comment:
I see this as a potential configuration:
- a property: `testcontainers.reuse.search.timeout`, and
- an env var: `TESTCONTAINERS_REUSE_SEARCH_TIMEOUT`

Reply:
Since the PR description says:
> We expect this change helps the community, but at the same time warn about its usage in parallel executions, as it could be the case that two concurrent test sessions get to the container creation at the same time, which could lead to the creation of two containers with the same request.

wouldn't we want to accept this limitation and allow for the race condition with the current implementation?

Reply:
Indeed. Added that comment as an idea for a potential follow-up